File: aoc24py/day03.py
# Advent of Code 2024
# Dr Lee A. Christie
#
# GitHub:   @leechristie
# Mastodon: @0x1ac@techhub.social
# Website:  leechristie.com


import time
import re


def get_file() -> str:
    all_lines = ''
    with open('input03.txt') as file:
        for line in file:
            all_lines += line
    return all_lines


def day03() -> None:

    start: float = time.perf_counter()

    part1: int = 0
    part2: int = 0

    all_lines = get_file()
    pattern = re.compile(r"mul\((\d\d?\d?),(\d\d?\d?)\)|do\(\)|don't\(\)")
    enabled = True
    for match in re.finditer(pattern, all_lines):
        match_string = all_lines[match.start():match.end()]
        if match_string == 'do()':
            enabled = True
        elif match_string == "don't()":
            enabled = False
        else:
            first, second = match.groups()
            value = int(first) * int(second)
            part1 += value
            if enabled:
                part2 += value


    stop: float = time.perf_counter()

    print("Advent of Code 2024")
    print("Day 3 - Mull It Over")
    print(f"Part 1: {part1}")
    print(f"Part 2: {part2}")
    print(f"Time Taken: {stop-start:.6f} s")

File: aoc24cpp/day03.cpp
// Advent of Code 2024
// Dr Lee A. Christie
//
// GitHub:   @leechristie
// Mastodon: @0x1ac@techhub.social
// Website:  leechristie.com

#include <fstream>
#include <iostream>
#include <regex>
#include <string>

#include "days.hpp"
#include "timing.hpp"

static std::string read_all(std::ifstream & file) {
    std::string all;
    std::string line;
    while (getline(file, line))
        all += line;
    return all;
}

void day03() {

    start_timer();

    int part1;
    int part2;
    {

        part1 = 0;
        part2 = 0;

        std::ifstream file {"input03.txt"};
        std::string all = read_all(file);

        std::regex pattern (R"(mul\((\d\d?\d?),(\d\d?\d?)\)|do\(\)|don't\(\))");

        bool enabled = true;

        const std::sregex_iterator end;
        for (auto iter = std::sregex_iterator(all.begin(), all.end(), pattern); iter != end; iter++) {

            const std::smatch & match = *iter;
            const std::string & str = match.str();

            // mul(###,###)
            if (str[0] == 'm') {

                const int product = stoi(match[1].str()) * stoi(match[2].str());
                part1 += product;
                if (enabled) {
                    part2 += product;
                }

            // do()
            } else if (str[2] == '(') {
                enabled = true;

            // don't()
            } else {
                enabled = false;
            }

        }

    }

    const double time = stop_timer();
    std::cout << "Advent of Code 2024" << std::endl;
    std::cout << "Day 3 - Mull It Over" << std::endl;
    std::cout << "Part 1: " << part1 << std::endl;
    std::cout << "Part 2: " << part2 << std::endl;
    std::cout << "Time Taken: " << std::fixed << std::setprecision(6) << time << " s" << std::endl;

}
